[#61178] Replace AuthCheckQuery with UserQuery.

ba1ash opened the pull request:
# Ticket
<!-- Provide the link to respective work package -->

<!-- Contributors: Please check our PR guide: https://www.openproject.org/docs/development/code-review-guidelines/#preparing-your-pull-request before opening a PR. -->

<!-- Reviewers: Please check our Review guide: https://www.openproject.org/docs/development/code-review-guidelines/#reviewing -->

https://community.openproject.org/work_packages/61178

# What are you trying to accomplish?
<!-- Provide a description of the changes. -->

I try to get a way to find out user name on storage side by only having an access token. OpenProject has to know user name on Nexcloud side to construct proper WebDAV paths, at least, and in OIDC scenario at the moment there is no way to have it.

# What approach did you choose and why?
<!-- This section is a place for you to describe your thought process in making these changes.
     List any tradeoffs you made to take on or pay down tech debt.
     Describe any alternative approaches you considered and why you discarded them. -->

I try to make AuthCheckQuery more "general". It has name UserQuery now and responds with additional data on success(id of the user on storage side). The query has been modified for both Nextcloud and OneDrive storage modules. Next step will be to use this query after token exchange to update RemoteIdentity.

# Merge checklist

- [x] Added/updated tests
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

ba1ash committed:
[#61178] Replace AuthCheckQuery with UserQuery.

https://community.openproject.org/work_packages/61178

NobodysNightmare commented on modules/storages/app/common/storages/peripherals/oauth_configurations/configuration_interface.rb:
```diff
@@ -41,6 +43,21 @@ def to_httpx_oauth_config = raise ::Storages::Errors::SubclassResponsibility
+            )
+        end
+
+        def storage_name = raise ::Storages::Errors::SubclassResponsibility
```
Defining this method might not be necessary. As far as I can tell, we always have a `@storage` when making these calls. At some point `Storage#to_s` was rewritten to return the provider name short form that can be used in registry resolve calls.

So what you should be able to do is `resolve("#{@storage}.your.stuff.here")`.

ba1ash replied:
Yes, true. That's absolutely unnecessary. Thanks!

NobodysNightmare commented on modules/storages/spec/common/storages/peripherals/storage_interaction/nextcloud/user_query_spec.rb:
```diff
@@ -0,0 +1,74 @@
+  let(:user) { create(:user) }
+  # let(:storage) do
+  #   create(:nextcloud_storage_with_local_connection, :as_not_automatically_managed, oauth_client_token_user: user)
+  # end
```
:yellow_circle: Before merging, we should remove outcommented code like this one (I guess a debugging left-over?).

NobodysNightmare reviewed (changes requested):
Small comments aside, this looks good. Nice to see that we can now treat Nextcloud and OneDrive so similar.

Pretty nice to imagine that the SSO case could effectively follow the same path, after further adjustments are done :)

mereghost commented on app/models/oauth_client.rb:
```diff
@@ -29,6 +31,7 @@
 class OAuthClient < ApplicationRecord
   belongs_to :integration, polymorphic: true
   has_many :remote_identities, dependent: :destroy
+  has_many :oauth_client_tokens, dependent: :destroy
```
Good catch. DB already has a ON DELETE CASCADE set anyway. =)

ba1ash replied:
Yes, I saw on db level. I added the statement just for convenience.

mereghost commented on modules/storages/app/common/storages/peripherals/oauth_configurations/nextcloud_configuration.rb:
```diff
@@ -47,8 +47,8 @@ def initialize(storage)
 
-        def extract_origin_user_id(rack_access_token)
-          rack_access_token.raw_attributes[:user_id]
+        def storage_name
+          "nextcloud"
         end
```
As per @NobodysNightmare comment, you can remote this. We made `Storage#to_s` to map to the `short_provider_name`. =)

mereghost commented on modules/storages/app/common/storages/peripherals/oauth_configurations/one_drive_configuration.rb:
```diff
@@ -51,11 +51,8 @@ def initialize(storage)
-            UrlBuilder.url(@storage.uri, "/v1.0/me"),
-            headers: { "Authorization" => "Bearer #{rack_access_token.access_token}", "Accept" => "application/json" }
-          ).raise_for_status.json["id"]
+        def storage_name
+          "one_drive"
         end
```
Same as previous comment

mereghost reviewed (commented):
Looks good so far. Just a couple of comments. =)

ba1ash committed:
[#61178] Fix specs and react on comments.

ba1ash committed:
[#61178] Downgrade httpx to 1.3.4.

Reason: htpx 1.4.0 gets mad in tests and hangs forever in combination with VCR and WebMock.

ba1ash committed:
[#61178] Fix one_drive/user_query_spec.rb.

ba1ash committed:
[#61178] Fix specs related to httpx downgrade.

NobodysNightmare commented on Gemfile:
```diff
@@ -235,7 +235,7 @@ gem "factory_bot_rails", "~> 6.4.4", require: false
 gem "turbo-rails", "~> 2.0.0"
 
-gem "httpx"
+gem "httpx", "~> 1.3.4"
```
IMO we should add a comment here, why we restrict the version and (probably much harder) we should try to create an upstream issue with reproduction instructions.

Did you get any idea how/why the newer version of HTTPX does that? Because it's working fine for most of our tests, so it surprises me that for one test it's showing this behaviour...

ba1ash replied:
I have no idea why yet. But it seems to be the case for all failures when VCR or WebMock stub can not be found. Then there is an infinite hang happening. I'll add a clarification comment.

NobodysNightmare replied:
The new comment is good, because it explains what to look out for to update.

In case you write/find an issue in their issue tracker, this would also be a great place to cross link it.

ba1ash replied:
Yes, sure, That is my plan.

ba1ash committed:
Merge branch 'dev' into implementation/61178-update-remoteidentity-after-token-exchange

ba1ash committed:
Try to fix manage_project_storage_spec.

ba1ash committed:
Merge remote-tracking branch 'origin/dev' into implementation/61178-update-remoteidentity-after-token-exchange

ba1ash committed:
[#61178] Add comment clarifying httpx version.

NobodysNightmare commented on spec/services/remote_identities/create_service_spec.rb:
```diff
@@ -4,23 +4,31 @@
 
-  subject(:service) { described_class.new(user:, oauth_config:, oauth_token:) }
+  before do
+    stub_request(:get, File.join(storage.host, "/ocs/v1.php/cloud/user")).to_return(status: 200, body: File.read("modules/storages/spec/support/payloads/nextcloud_user_query_success.xml"), headers: {})
```
Seeing that there are quite a bunch of line length offenses for essentially the same stubbed line. What do you think about offering a centralized helper to mock this?

E.g. `stub_nextcloud_user_query(storage.host)`

ba1ash replied:
Yes, good idea. Will do.

NobodysNightmare reviewed (approved):
Looks good to me. Approving already, but expecting that Rubocop is happy before merging ;-)

ba1ash committed:
Merge remote-tracking branch 'origin/dev' into implementation/61178-update-remoteidentity-after-token-exchange

ba1ash committed:
[#61178] Extract user_query stub to a method.
